src/testing/mock.rs

```diff
@@ -209,6 +226,91 @@ impl SpawnHandle for Spawner {
     }
 }
 
+#[derive(Clone)]
+pub struct IdleAwareSpawner {
+    spawner: Arc<Spawner>,
+    idle_mx: Arc<Mutex<()>>,
+    wake_flag: Arc<AtomicBool>,
+}
+
+struct SpawnFuture<T> {
+    task: Pin<Box<T>>,
+    wake_flag: Arc<AtomicBool>,
+}
+
+impl<T> SpawnFuture<T> {
+    fn new(task: T, wake_flag: Arc<AtomicBool>) -> Self {
+        SpawnFuture {
+            task: Box::pin(task),
+            wake_flag,
+        }
+    }
+}
+
+impl<T: Future<Output = ()> + Send + 'static> Future for SpawnFuture<T> {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.wake_flag
+            .store(false, std::sync::atomic::Ordering::Relaxed);
+        Future::poll(self.task.as_mut(), cx)
+    }
+}
+
+impl SpawnHandle for IdleAwareSpawner {
+    fn spawn(&self, _name: &'static str, task: impl Future<Output = ()> + Send + 'static) {
+        let wrapped = SpawnFuture::new(task, self.wake_flag.clone());
+        self.spawner.spawn(_name, wrapped)
+    }
+
+    fn spawn_essential(
+        &self,
+        name: &'static str,
+        task: impl Future<Output = ()> + Send + 'static,
+    ) -> TaskHandle {
+        let wrapped = SpawnFuture::new(task, self.wake_flag.clone());
+        self.spawner.spawn_essential(name, wrapped)
+    }
+}
+
+impl IdleAwareSpawner {
+    pub async fn wait(&self) {
+        self.spawner.wait().await
+    }
+
+    pub fn new() -> Self {
+        IdleAwareSpawner {
+            spawner: Arc::new(Spawner::new()),
+            idle_mx: Arc::new(Mutex::new(())),
+            wake_flag: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    pub async fn wait_idle(&self) {
+        let _ = self.idle_mx.lock();
```

Comment:
I don't see why this mutex is necessary. Also, if it is then I believe it doesn't work – if you assign it to `_` it immediately drops, you have to at least do `_x`.